src/guides/metamask-api/index.md

```diff
@@ -0,0 +1,1096 @@
+        ? JSON.parse(local)
+        : // backup if local storage is empty
+          { wallet: null, balance: null };
+
+      dispatch({ type: "pageLoaded", isMetamaskInstalled, wallet, balance });
+    }
+  }, []);
+
+  return (
+    <>
+      <Wallet />
+    </>
+  );
+};
+
+export default Home;
+```
+
+In this last page update to `pages/index.tsx`, we have relocated the `useEffect` from the `hooks/useMetaMask.tsx` page. This hook is consuming **dispatch**, so the proper next step would be to create a layout page with NextJS, but, since we only have one page, we simply added this code here.
+
+We have updated our `hooks/useMetamask.tsx` page's **PageLoadAction** to include **wallet** and **balance** as well as the code required to access our local storage and rehydrate our app.
+
+With those changes in place we have also updated our `components/Wallet.tsx` page to use our `useListen` hook since we are using that code in multiple places now, updated the `showAddToken` variable to a more descriptive name of `isConnected`, and added a `handleDisconnect()` function to dispatch an action clearing local storage in our browser.
+
+This also required a slight update to our JSX/HTML to display our buttons more neatly.
+
+[Checkout the Diff to see what changed](https://github.com/GuiBibeau/web3-unleashed-demo/pull/4/files)
+
+## Remove Listeners after Disconnect
+
+We have one final change we want to make to ensure that we stop listening to changes once the user has disconnected their wallet.
+
+We will update the `` and `` files. This will make TypeScript definitions file aware of the `removeAllListeners()` method we will be using, as well add necessary code to the `disconnect` case inside the `metamaskReducer`.
+
+Update `types.d.ts` file:
+
+```typescript
+type InjectedProviders = {
+  isMetaMask?: true;
+};
+
+interface Window {
+  ethereum: InjectedProviders & {
+    on: (...args: any[]) => void;
+    removeListener: (...args: any[]) => void;
+    removeAllListeners: (...args: any[]) => void;
+    request<T = any>(args: any): Promise<T>;
+  };
+}
+```
+
+Update the case statement in the `useMetamask.tsx` file to:
+
+```typescript
+    case "disconnect": {
+      window.localStorage.removeItem("metamaskState");
+      if (typeof window.ethereum !== undefined) {
+        window.ethereum.removeAllListeners(["accountsChanged"]);
+      }
+      return { ...state, wallet: null, balance: null };
+    }
+```
+
+Again, here we have ensured that all listeners added after connecting the wallet stop listening once the user is disconnected.
+
+This concludes the demo! But you're just getting started; for a challenge, try updating the UI, try to add functionality to switch chains, and overall, have fun. If you have any questions or need help with MetaMask, reach out to our DevRel team on Twitter. You can contact [Gui Bibeau](https://twitter.com/guibibeau) and [Eric Bishard](https://twitter.com/httpjunkie) with any questions or feedback.
```

Comment:
What do you think about including a reference to the [final](https://github.com/GuiBibeau/Metamask-Nextjs-Starter/tree/final) branch so users can optionally switch to it in case they're having issues following along?

Reply:
Good suggestion.